mhcflurry/dataset.py

```diff
@@ -0,0 +1,554 @@
+# Copyright (c) 2016. Mount Sinai School of Medicine
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import (
+    print_function,
+    division,
+    absolute_import,
+)
+from collections import defaultdict, OrderedDict
+import logging
+
+
+from six import string_types
+import pandas as pd
+import numpy as np
+
+from .common import geometric_mean
+from .dataset_helpers import (
+    prepare_pMHC_affinity_arrays,
+    load_dataframe
+)
+from .peptide_encoding import fixed_length_index_encoding
+from .imputation_helpers import (
+    check_dense_pMHC_array,
+    prune_dense_matrix_and_labels,
+    dense_pMHC_matrix_to_nested_dict,
+)
+
+
+class Dataset(object):
+
+    """
+    Peptide-MHC binding dataset with helper methods for constructing
+    different representations (arrays, DataFrames, dictionaries, &c).
+
+    Design considerations:
+        - want to allow multiple measurements for each pMHC pair (which can
+          be dynamically combined)
+        - optional sample weights associated with each pMHC measurement
+    """
+    def __init__(self, df):
+        """
+        Constructs a DataSet from a pandas DataFrame with the following
+        columns:
+            - allele
+            - peptide
+            - affinity
+
+        Also, there is an optional column:
+            - sample_weight
+
+        If `sample_weight` is missing then it is filled with a default value
+        of 1.0
+
+        Parameters
+        ----------
+        df : pandas.DataFrame
+        """
+        columns = set(df.columns)
+
+        for expected_column_name in {"allele", "peptide", "affinity"}:
+            if expected_column_name not in columns:
+                raise ValueError("Missing column '%s' from DataFrame")
+        # make allele and peptide columns the index, and copy it
+        # so we can add a column without any observable side-effect in
+        # the calling code
+        df = df.set_index(["allele", "peptide"], drop=False)
+
+        if "sample_weight" not in columns:
+            df["sample_weight"] = np.ones(len(df), dtype=float)
```

Comment:
nothing wrong with what you have, but just noting it could also be `df["sample_weight"] = 1.0`